File: sample-videochat-webrtc/sample-videochat-webrtc/Classes/QMSoundManager/QMSoundManager.m
//
//  QMSoundManager.m
//  Qmunicate
//
//  Created by Andrey Ivanov on 01.07.14.
//  Copyright (c) 2014 Quickblox. All rights reserved.
//

#import "QMSoundManager.h"

static NSString * const kSystemSoundTypeCAF = @"caf";
static NSString * const kSystemSoundTypeAIF = @"aif";
static NSString * const kSystemSoundTypeAIFF = @"aiff";
static NSString * const kystemSoundTypeWAV = @"wav";

static NSString * const kQMSoundManagerSettingKey = @"kQMSoundManagerSettingKey";

@interface QMSoundManager() {
    
    NSMutableDictionary *_sounds;
    NSMutableDictionary *_completionBlocks;
    BOOL _audioDeviceChanged;
}

@end

@implementation QMSoundManager

- (void)dealloc {
    
    NSNotificationCenter *notifcationCenter =
    [NSNotificationCenter defaultCenter];
    [notifcationCenter removeObserver:self];
}

void systemServicesSoundCompletion(SystemSoundID  soundID, void *data) {
    
    void(^completion)(void) = [QMSoundManager.instance completionBlockForSoundID:soundID];
    
    if (completion) {
        
        completion();
        [QMSoundManager.instance  removeCompletionBlockForSoundID:soundID];
    }
}

+ (instancetype)instance {
    
    static id instance;
    
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        
        instance = [[self alloc] init];
    });
    
    return instance;
}

- (instancetype)init {
    
    self = [super init];
    if (self) {
        
        self.on = YES;
        
        _sounds = [NSMutableDictionary dictionary];
        _completionBlocks = [NSMutableDictionary dictionary];
        
        NSNotificationCenter *notifcationCenter =
        [NSNotificationCenter defaultCenter];
        
        [notifcationCenter addObserver:self
                              selector:@selector(didReceiveMemoryWarningNotification:)
                                  name:UIApplicationDidReceiveMemoryWarningNotification
                                object:nil];
    }
    
    return self;
}

- (void)setOn:(BOOL)on {
    
    _on = on;
    
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    [userDefaults setBool:on forKey:kQMSoundManagerSettingKey];
    [userDefaults synchronize];
    
    if (!on) {
        
        [self stopAllSounds];
    }
}

#pragma mark - Playing sounds

- (void)playSoundWithName:(NSString *)filename
                extension:(NSString *)extension
                  isAlert:(BOOL)isAlert
               completion:(void(^)(void))completion {
    
    if (!self.on) {
        return;
    }
    
    if (!filename || !extension) {
        return;
    }
    
    if (!_sounds[filename]) {
        
        [self addSoundIDForAudioFileWithName:filename
                                   extension:extension];
    }
    
    SystemSoundID soundID = [self soundIDForFilename:filename];
    
    if (soundID) {
        
        if (completion) {
            
            OSStatus error =
            AudioServicesAddSystemSoundCompletion(soundID,
                                                  NULL,
                                                  NULL,
                                                  systemServicesSoundCompletion,
                                                  NULL);
            if (error) {
                
                [self logError:error
                   withMessage:@"Warning! Completion block could not be added to SystemSoundID."];
            }
            else {
                
                [self addCompletionBlock:completion
                               toSoundID:soundID];
            }
        }
        
        if (isAlert) {
            AudioServicesPlayAlertSound(soundID);
        }
        else {
            AudioServicesPlaySystemSound(soundID);
        }
    }
}

- (void)playSoundWithName:(NSString *)filename extension:(NSString *)extension {
    
    [self playSoundWithName:filename
                  extension:extension
                 completion:nil];
}

- (void)playSoundWithName:(NSString *)filename
                extension:(NSString *)extension
               completion:(void(^)(void))completion {
    
    [self playSoundWithName:filename
                  extension:extension
                    isAlert:NO
                 completion:completion];
}

- (void)playAlertSoundWithName:(NSString *)filename
                     extension:(NSString *)extension
                    completion:(void(^)(void))completion {
    
    [self playSoundWithName:filename
                  extension:extension
                    isAlert:YES
                 completion:completion];
}

- (void)playAlertSoundWithName:(NSString *)filename
                     extension:(NSString *)extension {
    
    [self playAlertSoundWithName:filename
                       extension:extension
                      completion:nil];
}

- (void)playVibrateSound {
    
    if (self.on) {
        
        AudioServicesPlaySystemSound(kSystemSoundID_Vibrate);
    }
}

- (void)stopAllSounds {
    [self unloadSoundIDs];
}

- (void)stopSoundWithFilename:(NSString *)filename {
    
    SystemSoundID soundID = [self soundIDForFilename:filename];
    NSData *data = [self dataWithSoundID:soundID];
    
    [self unloadSoundIDForFileNamed:filename];
    
    [_sounds removeObjectForKey:filename];
    [_completionBlocks removeObjectForKey:data];
}

- (void)preloadSoundWithFilename:(NSString *)filename
                       extension:(NSString *)extension {
    
    if (!_sounds[filename]) {
        [self addSoundIDForAudioFileWithName:filename
                                   extension:extension];
    }
}

#pragma mark - Sound data

- (NSData *)dataWithSoundID:(SystemSoundID)soundID {
    
    return [NSData dataWithBytes:&soundID
                          length:sizeof(SystemSoundID)];
}

- (SystemSoundID)soundIDFromData:(NSData *)data {
    
    if (data) {
        
        SystemSoundID soundID;
        [data getBytes:&soundID length:sizeof(SystemSoundID)];
        return soundID;
    }
    
    return 0;
}

#pragma mark - Sound files

- (SystemSoundID)soundIDForFilename:(NSString *)filenameKey {
    
    NSData *soundData = _sounds[filenameKey];
    return [self soundIDFromData:soundData];
}

- (void)addSoundIDForAudioFileWithName:(NSString *)filename
                             extension:(NSString *)extension {
    
    SystemSoundID soundID = [self createSoundIDWithName:filename
                                              extension:extension];
    if (soundID) {
        
        NSData *data = [self dataWithSoundID:soundID];
        _sounds[filename] = data;
    }
}

#pragma mark - Sound completion blocks

- (void(^)(void))completionBlockForSoundID:(SystemSoundID)soundID {
    
    NSData *data = [self dataWithSoundID:soundID];
    return _completionBlocks[data];
}

- (void)addCompletionBlock:(void(^)(void))block
                 toSoundID:(SystemSoundID)soundID {
    
    NSData *data = [self dataWithSoundID:soundID];
    _completionBlocks[data] = [block copy];
}

- (void)removeCompletionBlockForSoundID:(SystemSoundID)soundID {
    
    NSData *key = [self dataWithSoundID:soundID];
    [_completionBlocks removeObjectForKey:key];
    AudioServicesRemoveSystemSoundCompletion(soundID);
}

#pragma mark - Managing sounds

- (SystemSoundID)createSoundIDWithName:(NSString *)filename
                             extension:(NSString *)extension {
    
    NSURL *fileURL = [[NSBundle mainBundle] URLForResource:filename
                                             withExtension:extension];
    
    if ([[NSFileManager defaultManager] fileExistsAtPath:[fileURL path]]) {
        
        SystemSoundID soundID;
        OSStatus error = AudioServicesCreateSystemSoundID((__bridge CFURLRef)fileURL, &soundID);
        
        if (error) {
            [self logError:error withMessage:@"Warning! SystemSoundID could not be created."];
            return 0;
        }
        else {
            return soundID;
        }
    }
    
    NSLog(@"Error: audio file not found at URL: %@", fileURL);
    
    return 0;
}

- (void)unloadSoundIDs {
    
    for(NSString *eachFilename in [_sounds allKeys]) {
        [self unloadSoundIDForFileNamed:eachFilename];
    }
    
    [_sounds removeAllObjects];
    [_completionBlocks removeAllObjects];
}

- (void)unloadSoundIDForFileNamed:(NSString *)filename {
    
    SystemSoundID soundID = [self soundIDForFilename:filename];
    
    if(soundID) {
        AudioServicesRemoveSystemSoundCompletion(soundID);
        
        OSStatus error = AudioServicesDisposeSystemSoundID(soundID);
        
        if(error) {
            
            [self logError:error withMessage:@"Warning! SystemSoundID could not be disposed."];
        }
    }
}

- (void)logError:(OSStatus)error withMessage:(NSString *)message {
    
    NSString *errorMessage = nil;
    
    switch (error) {
            
        case kAudioServicesUnsupportedPropertyError: errorMessage = @"The property is not supported."; break;
        case kAudioServicesBadPropertySizeError: errorMessage = @"The size of the property data was not correct."; break;
        case kAudioServicesBadSpecifierSizeError: errorMessage = @"The size of the specifier data was not correct."; break;
        case kAudioServicesSystemSoundUnspecifiedError:errorMessage = @"An unspecified error has occurred."; break;
        case kAudioServicesSystemSoundClientTimedOutError: errorMessage = @"System sound client message timed out."; break;
    }
    
    NSLog(@"%@ Error: (code %d) %@", message, (int)error, errorMessage);
}

#pragma mark - Did Receive Memory Warning Notification

- (void)didReceiveMemoryWarningNotification:(NSNotification *)notification {
    
    [self unloadSoundIDs];
}

#pragma mark - Default sounds

NSString *const kQMCallingSoundName = @"calling";
NSString *const kQMBusySoundName = @"busy";
NSString *const kQMEndOfCallSoundName = @"end_of_call";
NSString *const kQMRingtoneSoundName = @"ringtone";

+ (void)playCallingSound {
    
    [[QMSoundManager instance] playSoundWithName:kQMCallingSoundName
                                       extension:kystemSoundTypeWAV];
}

+ (void)playBusySound {
    
    [[QMSoundManager instance] playSoundWithName:kQMBusySoundName
                                       extension:kystemSoundTypeWAV];
}

+ (void)playEndOfCallSound {
    
    [[QMSoundManager instance] playSoundWithName:kQMEndOfCallSoundName
                                       extension:kystemSoundTypeWAV];
}

+ (void)playRingtoneSound {
    
    [[QMSoundManager instance] playAlertSoundWithName:kQMRingtoneSoundName
                                            extension:kystemSoundTypeWAV];
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/sample-conference-videochat-swift/Components/QMSoundManager/QMSoundManager.swift
//
//  QMSoundManager.swift
//  sample-conference-videochat-swift
//
//  Created by Vladimir Nybozhinsky on 04.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import Foundation
import AVFoundation
import UIKit

struct QMSoundManagerConstant {
    static let caf = "caf"
    static let aif = "aif"
    static let aiff = "aiff"
    static let wav = "wav"
    
    static let soundManagerSettingKey = "kQMSoundManagerSettingKey"
    
    static let calling = "calling"
    static let busy = "busy"
    static let endOfCall = "end_of_call"
    static let ringtone = "ringtone"
}

struct SoundErrorConstant {
    static let notSupported = NSLocalizedString("The property is not supported.", comment: "")
    static let sizePropertyDataNotCorrect = NSLocalizedString("The size of the property data was not correct.", comment: "")
    static let sizeSpecifierDataNotCorrect = NSLocalizedString("The size of the specifier data was not correct.", comment: "")
    static let unspecifiedErrorOccurred = NSLocalizedString("An unspecified error has occurred.", comment: "")
    static let timeOut = NSLocalizedString("System sound client message timed out.", comment: "")
    static let notBeAdded = NSLocalizedString("Warning! Completion block could not be added to SystemSoundID.", comment: "")
    static let notBeCreated = NSLocalizedString("Warning! SystemSoundID could not be created.", comment: "")
    static let notBeDisposed = NSLocalizedString("Warning! SystemSoundID could not be disposed.", comment: "")
}

class QMSoundManager {
    var on = false
    
    //MARK: Instance Singelton
    static let instance = QMSoundManager()
    
    //MARK: Variadles
    private var sounds: [String : Data] = [:]
    private var completionBlocks: [AnyHashable : Any] = [:]
    private var audioDeviceChanged = false
    
    let notifcationCenter = NotificationCenter.default
    
    // MARK: Life cycle
    
    init() {
        on = true
        notifcationCenter.addObserver(self, selector: #selector(self.didReceiveMemoryWarning(_:)), name: UIApplication.didReceiveMemoryWarningNotification, object: nil)
        
    }

    deinit {
        notifcationCenter.removeObserver(self)
    }
    
    func setOn(_ on: Bool) {
        self.on = on
        
        let userDefaults = UserDefaults.standard
        userDefaults.set(on, forKey: QMSoundManagerConstant.soundManagerSettingKey)
        userDefaults.synchronize()
        
        if !on {
            stopAllSounds()
        }
    }
    
    /**
     *  Plays a system sound object corresponding to an audio file with the given filename and extension.
     *  The system sound player will lazily initialize and load the file before playing it, and then cache its corresponding `SystemSoundID`.
     *  If this file has previously been played, it will be loaded from cache and played immediately.
     *
     *  @param filename      A string containing the base name of the audio file to play.
     *  @param extension A string containing the extension of the audio file to play.
     *  This parameter must be one of `caf`, `aif`, `aiff`, or `wav`
     *
     *  @warning If the system sound object cannot be created, this method does nothing.
     */
    func playSound(withName filename: String, ext: String?) {
        playSound(withName: filename, ext: ext)
    }
    
    func playSound(withName filename: String, ext: String?, isAlert: Bool, completion: @escaping () -> Void) {
        if on == false {
            return
        }
        
        if filename.count > 0 || ext == nil {
            return
        }
        
        if sounds[filename] == nil {
            
            addSoundIDForAudioFile(withName: filename, ext: ext)
        }
        
        let soundID = self.soundID(forFilename: filename)

        let weakSelf = UnsafeMutableRawPointer(Unmanaged.passUnretained(self).toOpaque())
        let error = AudioServicesAddSystemSoundCompletion(soundID,
                                                          nil,
                                                          nil,
                                                          {soundID, weakSelfPointer in
                                                            
                                                            guard let weakSelfPointer = weakSelfPointer else {
                                                                return
                                                            }
                                                            
                                                            let weakSelfValue = Unmanaged<QMSoundManager>.fromOpaque(weakSelfPointer).takeUnretainedValue()
                                                            
                                                            // Then you can use `weakSelfValue` as you would do with `self`.
                                                            weakSelfValue.systemServicesSoundCompletion(soundID: soundID, data: weakSelfPointer)
        }, weakSelf)
            
            if error != noErr {
                
                logError(error, withMessage: SoundErrorConstant.notBeAdded)
            } else {
                addCompletionBlock(completion, to: soundID)
            }
            
            if isAlert {
                AudioServicesPlayAlertSound(soundID)
            } else {
                AudioServicesPlaySystemSound(soundID)
            }
    }
    
    /**
     *  Plays a system sound object *as an alert* corresponding to an audio file with the given filename and extension.
     *  The system sound player will lazily initialize and load the file before playing it, and then cache its corresponding `SystemSoundID`.
     *  If this file has previously been played, it will be loaded from cache and played immediately.
     *
     *  @param filename     A string containing the base name of the audio file to play.
     *  @param extension    A string containing the extension of the audio file to play.
     *  This parameter must be one of `caf`, `aif`, `aiff`, or `wav
     *
     *  @warning If the system sound object cannot be created, this method does nothing.
     *
     *  @warning This method performs the same functions as `playSoundWithName: extension:`, with the excepion that,
     *  depending on the particular iOS device, this method may invoke vibration.
     */
    func playAlertSound(withName filename: String, ext: String?) {
        playAlertSound(withName: filename, ext: ext)
    }
    // MARK: Preset sounds
    class func playCallingSound() {
        QMSoundManager.instance.playSound(withName: QMSoundManagerConstant.calling, ext: QMSoundManagerConstant.wav)
    }
    
    class func playBusySound() {
        QMSoundManager.instance.playSound(withName: QMSoundManagerConstant.busy, ext: QMSoundManagerConstant.wav)
    }
    
    class func playEndOfCallSound() {
        QMSoundManager.instance.playSound(withName: QMSoundManagerConstant.endOfCall, ext: QMSoundManagerConstant.wav)
    }
    
    class func playRingtoneSound() {
        QMSoundManager.instance.playAlertSound(withName: QMSoundManagerConstant.ringtone, ext: QMSoundManagerConstant.wav)
    }
    
    func playAlertSound(withName filename: String, ext: String?, completion: @escaping () -> Void) {
        playSound(withName: filename, ext: ext, isAlert: true, completion: completion)
    }
    
    func playSound(withName filename: String, ext: String?, completion: @escaping () -> Void) {
        playSound(withName: filename, ext: ext, isAlert: false, completion: completion)
    }
    
    /**
     *  On some iOS devices, you can call this method to invoke vibration.
     *  On other iOS devices this functionaly is not available, and calling this method does nothing.
     */
    func playVibrateSound() {
        if on == true {
            AudioServicesPlaySystemSound(kSystemSoundID_Vibrate)
        }
    }
    
    /**
     *  Stops playing all sounds immediately.
     *
     *  @warning Any completion blocks attached to any currently playing sound will *not* be executed.
     *  Also, calling this method will purge all `SystemSoundID` objects from cache, regardless of whether or not they were currently playing.
     */
    func stopAllSounds() {
        unloadSoundIDs()
    }
    
    /**
     *  Stops playing the sound with the given filename immediately.
     *
     *  @param filename The filename of the sound to stop playing.
     *
     *  @warning If a completion block is attached to the given sound, it will *not* be executed.
     *  Also, calling this method will purge the `SystemSoundID` object for this file from cache, regardless of whether or not it was currently playing.
     */
    func stopSound(withFilename filename: String) {
        
        let soundID = self.soundID(forFilename: filename)
        let data: Data? = self.data(with: soundID)
        
        unloadSoundID(forFileNamed: filename)
        
        sounds.removeValue(forKey: filename)
        completionBlocks.removeValue(forKey: data)
    }
    
    /**
     *  Preloads a system sound object corresponding to an audio file with the given filename and extension.
     *  The system sound player will initialize, load, and cache the corresponding `SystemSoundID`.
     *
     *  @param filename      A string containing the base name of the audio file to play.
     *  @param extension A string containing the extension of the audio file to play.
     *  This parameter must be one of `caf`, `aif`, `aiff`, or `wav`.
     */
    func preloadSound(withFilename filename: String, ext: String?) {
        if sounds[filename] == nil {
            addSoundIDForAudioFile(withName: filename, ext: ext)
        }
    }
    
    func systemServicesSoundCompletion(soundID: SystemSoundID, data: UnsafeMutableRawPointer?) {
        let completion: (() -> Void)? = QMSoundManager.instance.completionBlock(for: soundID)
        if completion != nil {
            completion?()
            QMSoundManager.instance.removeCompletionBlock(for: soundID)
        }
    }
    
    // MARK: - Sound completion blocks
    func completionBlock(for soundID: SystemSoundID) -> (() -> Void) {
        let data: Data? = self.data(with: soundID)
        if let aData = data {
            return completionBlocks[aData] as! (() -> Void)
        }
        return {  }
    }
    
    func addCompletionBlock(_ block: @escaping () -> Void, to soundID: SystemSoundID) {
        let data: Data? = self.data(with: soundID)
        if let soundData = data {
            completionBlocks[soundData] = block()
        }
    }
    
    func removeCompletionBlock(for soundID: SystemSoundID) {
        let key: Data? = data(with: soundID)
        completionBlocks.removeValue(forKey: key)
        AudioServicesRemoveSystemSoundCompletion(soundID)
    }
    
    // MARK: - Sound data
    func data(with soundID: SystemSoundID) -> Data? {
        var soundID = soundID
        let data = Data(bytes:  &soundID, count: MemoryLayout<SystemSoundID>.size)
        return data
    }
    
    func soundID(from data: Data?) -> SystemSoundID {
        if data != nil {
            var soundID: UInt8 = 0
            data?.copyBytes(to: &soundID, count: MemoryLayout<SystemSoundID>.size)
            return UInt32(soundID) as SystemSoundID
        }
        return SystemSoundID(0)
    }
    
    // MARK: - Sound files
    func soundID(forFilename filenameKey: String) -> SystemSoundID {
        let soundData: Data? = sounds[filenameKey]
        return soundID(from: soundData)
    }
    
    func addSoundIDForAudioFile(withName filename: String, ext: String?) {
        let soundID = createSoundID(withName: filename, ext: ext)
        if soundID != 0 && filename.count > 0 {
            let data: Data? = self.data(with: soundID)
            sounds[filename] = data
        }
    }
    
    // MARK: - Managing sounds
    func createSoundID(withName filename: String, ext: String?) -> SystemSoundID {
        let fileURL: URL? = Bundle.main.url(forResource: filename, withExtension: ext)
        if FileManager.default.fileExists(atPath: fileURL?.path ?? "") {
            var soundID: SystemSoundID = 0
            let error: OSStatus = AudioServicesCreateSystemSoundID(fileURL! as CFURL, &soundID)
            
            if error != noErr  {
                logError(error, withMessage: SoundErrorConstant.notBeCreated)
                return SystemSoundID(0)
            } else {
                return soundID
            }
        }
        
        if let fileURL = fileURL {
            print("Error: audio file not found at URL: \(fileURL)")
        }
        return SystemSoundID(0)
    }
    
    func unloadSoundIDs() {
        for fileName in sounds.keys {
            unloadSoundID(forFileNamed: fileName)
        }
        sounds.removeAll()
        completionBlocks.removeAll()
    }
    
    func unloadSoundID(forFileNamed filename: String) {
        let soundID = self.soundID(forFilename: filename)
        if soundID != 0 {
            AudioServicesRemoveSystemSoundCompletion(soundID)
            let error: OSStatus = AudioServicesDisposeSystemSoundID(soundID)
            if error != noErr  {
                logError(error, withMessage: SoundErrorConstant.notBeDisposed)
            }
        }
    }
    
    func logError(_ error: OSStatus, withMessage message: String?) {
        var errorMessage: String? = nil
        
        switch error {
        case kAudioServicesUnsupportedPropertyError:
            errorMessage = SoundErrorConstant.notSupported
        case kAudioServicesBadPropertySizeError:
            errorMessage = SoundErrorConstant.sizePropertyDataNotCorrect
        case kAudioServicesBadSpecifierSizeError:
            errorMessage = SoundErrorConstant.sizeSpecifierDataNotCorrect
        case kAudioServicesSystemSoundUnspecifiedError:
            errorMessage = SoundErrorConstant.unspecifiedErrorOccurred
        case kAudioServicesSystemSoundClientTimedOutError:
            errorMessage = SoundErrorConstant.timeOut
        default:
            break
        }
        print("\(message ?? "") Error: (code \(Int(error))) \(errorMessage ?? "")")
    }
    
    // MARK: - Did Receive Memory Warning Notification
    @objc func didReceiveMemoryWarning(_ notification: Notification?) {
        unloadSoundIDs()
    }
}
